README.md
# time2backup server
A server part to allow time2backup clients save files on an ssh server.

## Download and install
1. Download the last release of time2backup server [here](https://github.com/time2backup/server/releases)
2. Uncompress archive where you want
3. Edit the config file

## Documentation
For global usage, see the [user manual](docs/user_manual.md).

## License
time2backup server is licensed under the MIT License. See [LICENSE.md](LICENSE.md) for the full license text.

## Credits
Author: Jean Prunneaux https://jean.prunneaux.com

Source code: https://github.com/time2backup/server


CHANGELOG.md
# Changelog

## 1.0.3 (2021-03-04)
- Minor code improvements
- Upgrade libbash.sh to 1.20.0

## 1.0.2 (2021-01-29)
- Upgrade libbash.sh to 1.18.1

## 1.0.1 (2020-08-24)
- Upgrade libbash.sh to 1.17.1, improving import config security
- Minor code improvements

## 1.0.0 (2019-11-23)
- Minor code improvements
- Removed root security

## 1.0.0 RC 3 (2019-11-01)
- Add install/uninstall scripts
- Removed sudo mode
- Upgrade libbash.sh 1.14.4

## 1.0.0 RC 2 (2019-10-30)
- Changed exit codes
- Secured config import

## 1.0.0 RC 1 (2019-10-21)
- Restore command support
- Trash mode support

## 1.0.0 beta 2 (2019-10-18)
- Trash mode support
- Improve logs and display

## 1.0.0 beta 1 (2019-10-17)
- First release with backup command support

---------------------------------------------------------------

## License
time2backup server is licensed under the MIT License. See [LICENSE.md](LICENSE.md) for the full license text.

## Credits
Author: Jean Prunneaux  https://jean.prunneaux.com

Website: https://github.com/time2backup/server


install.sh
#!/bin/bash
#
#  time2backup server install script
#
#  MIT License
#  Copyright (c) 2017-2021 Jean Prunneaux
#  Website: https://github.com/time2backup/server
#

if [ "$(whoami)" != root ] ; then
	echo "You must be root to install time2backup server!"
	exit 1
fi

# current directory
curdir=$(dirname "$0")

# load libbash
source "$curdir"/libbash/libbash.sh - &> /dev/null
if [ $? != 0 ] ; then
	echo >&2 "internal error"
	exit 1
fi

# create global link
ln -sf "$lb_current_script_directory/t2b-server.sh" /usr/bin/time2backup-server
if [ $? != 0 ] ; then
	lb_error "Cannot create global link"
	exit 1
fi

# create t2b group if not exists
if ! grep -q '^t2b-server:' /etc/group ; then
	addgroup t2b-server
	if [ $? != 0 ] ; then
		lb_error "Cannot create group t2b-server"
		exit 1
	fi
fi

# secure current directory permissions
chmod -R 750 "$curdir"
chmod -x "$curdir"/*.md "$curdir"/config/* "$curdir"/docs/* "$curdir"/inc/* \
         "$curdir"/libbash/*.* "$curdir"/libbash/*/*
touch "$curdir"/.access && chmod 660 "$curdir"/.access
touch "$curdir"/config/auth.conf && chmod 640 "$curdir"/config/auth.conf
touch "$curdir"/server.log && chmod 660 "$curdir"/server.log

[ -f "$curdir"/config/time2backup-server.conf ] || cp "$curdir"/config/time2backup-server.example.conf "$curdir"/config/time2backup-server.conf
chmod 640 "$curdir"/config/time2backup-server.conf

chown -R root:t2b-server "$curdir"


inc/commands.sh
#
#  time2backup server commands
#
#  This file is part of time2backup server (https://github.com/time2backup/server)
#
#  MIT License
#  Copyright (c) 2017-2021 Jean Prunneaux
#

# Index
#
#   Commands
#     t2bs_prepare
#     t2bs_backup
#     t2bs_restore
#     t2bs_history
#     t2bs_rotate
#     t2bs_rsync


# Initialize connection and prepare for backup process
# Usage: t2bs_prepare COMMAND [OPTIONS]
t2bs_prepare() {

	local cmd=$1 resume=false force_unlock=false args=()
	shift

	# prepare destination
	if ! prepare_destination &> /dev/null ; then
		print_error --log "destination not writable"
		return 204
	fi

	# prepare server response
	echo "[server]"

	case $cmd in
		backup)
			# Usage: backup [OPTIONS] DATE PATH
			while [ $# -gt 0 ] ; do
				case $1 in
					--resume)
						resume=true
						;;
					--unlock)
						force_unlock=true
						;;
					*)
						break
						;;
				esac
				shift
			done

			backup_date=$1
			shift

			# test backup date format
			if ! check_backup_date "$backup_date" ; then
				usage_error "prepare backup: backup date format not conform: $backup_date"
				return 201
			fi

			src=$*

			# test if path is defined
			if [ -z "$src" ] ; then
				usage_error "prepare restore: source is missing"
				return 201
			fi

			# test if a backup is running
			if current_lock -q ; then
				# force mode: delete old lock
				if $force_unlock ; then
					if ! release_lock -f ; then
						print_error --log "failed to release lock"
						return 202
					fi
				else
					print_error "backup is already running"
					return 205
				fi
			fi

			# forward args to write in credentials file
			args=(date="$backup_date" src="$src")
			$resume && args+=(resume=true)

			# get latest backup for source
			local first=true b history=($(get_backup_history -a -n "$src"))
			for b in "${history[@]}" ; do
				# resume from last backup: skip latest backup
				if $first ; then
					if $resume ; then
						first=false
						continue
					else
						# search if last backup needs to be resumed

						# check status of the last backup
						if ! check_infofile_rsync_result $b "$src" ; then
							args+=(resume=true)
							log_debug "prepare: next backup will be resumed from $b"
							first=false
							continue
						fi
					fi
				fi

				# get trash if different from actual date
				if [ "$b" != "$1" ] ; then
					echo "trash = $b"
					args+=(trash="$b")
					break
				fi

				first=false
			done
			;;

		restore)
			# Usage: restore DATE PATH

			backup_date=$1
			shift

			# test backup date format
			if ! check_backup_date "$backup_date" ; then
				usage_error "prepare restore: backup date format not conform: $backup_date"
				return 201
			fi

			# test if path is defined
			if [ -z "$*" ] ; then
				usage_error "prepare restore: source is missing"
				return 201
			fi

			debug "search versions for $*"

			# get backup versions of this file
			file_history=($(get_backup_history "$*"))

			# recheck if date is still there
			local found=false
			if [ ${#file_history[@]} -gt 0 ] ; then
				lb_in_array "$backup_date" "${file_history[@]}" && found=true
			fi

			if ! $found ; then
				print_error "Backup version has vanished! Please retry later"
				return 202
			fi

			# test if latest is a directory
			local backup_path=$(get_backup_path "$*")

			echo "rsync_result = $(get_infofile_value "$destination/$backup_date/backup.info" "$*" rsync_result)"

			[ -n "$backup_path" ] && [ -d "$destination/$backup_date/$backup_path" ] && \
				echo "src_type = directory"

			# check if a current backup is running
			current_lock -q && echo "status = running"

			# forward args to write in credentials file
			args=(date=$backup_date)
			;;

		*)
			# bad command
			usage_error "bad command: $*"
			return 201
			;;
	esac

	# generate token
	token=$(lb_generate_password 32 | tr -cd '[:alnum:]')
	token=${token:0:16}

	# write token in credentials file
	if ! srv_save_token "$token" "${args[@]}" ; then
		internal_error "write token failed"
		return 202
	fi

	# output infos
	echo "version = $version"
	echo "token = $token"
	echo "destination = \"$destination\""
	echo "hard_links = $hard_links"
}


# Perform backup
# Usage: t2bs_backup [OPTIONS] RSYNC_OPTIONS
t2bs_backup() {

	# get command options
	while [ $# -gt 0 ] ; do
		case $1 in
			--t2b-rotate)
				if [ -z "$2" ] ; then
					usage_error "backup: --t2b-rotate argument missing"
					return 201
				fi
				keep_limit=$2
				shift
				;;

			--t2b-keep)
				if ! lb_is_integer $2 ; then
					usage_error "backup: --t2b-keep argument not conform: $2"
					return 201
				fi
				clean_keep=$2
				shift
				;;

			--t2b-trash)
				trash_mode=true
				;;

			*)
				break
				;;
		esac
		shift # load next argument
	done

	# check rsync syntax
	srv_check_rsync_options "$@" || return 201

	# get infos from token
	for i in "${token_infos[@]}" ; do
		case $(echo "$i" | cut -d= -f1) in
			date)
				backup_date=$(echo "$i" | cut -d= -f2-)
				;;
			src)
				src=$(echo "$i" | cut -d= -f2-)
				;;
			trash)
				last_clean_backup=$(echo "$i" | cut -d= -f2-)
				;;
			resume)
				resume=$(echo "$i" | cut -d= -f2-)
				;;
		esac
	done

	# verify if required variables are there
	if [ -z "$backup_date" ] || [ -z "$src" ] ; then
		internal_error "failed to get infos from token"
		return 202
	fi

	# check lock file
	lock_file=$(current_lock -f)
	if [ -n "$lock_file" ] ; then
		# lock file exists: compare token
		if [ "$(lb_get_config "$lock_file" token)" != "$token" ] ; then
			print_error "destination locked"
			log_error "token provided valid, but different from lock file for user $lb_current_user $ssh_info"
			return 205
		fi
	else
		# create lock
		create_lock &> /dev/null && \
		lb_set_config "$(current_lock -f)" token "$token"
	fi

	# catch term signals (must be silent for some OS)
	catch_kills srv_cancel_exit &> /dev/null

	create_infofile

	# get source path
	path_dest=$(get_backup_path "$src")

	# write new source section to info file
	lb_set_config -s src1 "$infofile" path "$src"
	lb_set_config -s src1 "$infofile" rsync_result -1
	lb_set_config -s src1 "$infofile" trash $last_clean_backup

	# resume: search the last backup
	if lb_istrue $resume ; then
		local b history=($(get_backup_history -n -a "$src")) resume_date

		for b in "${history[@]}" ; do
			# mirror mode: take the latest
			if [ $keep_limit = 0 ] ; then
				resume_date=$b
				break
			fi

			# ignore the last clean backup
			if [ "$b" != "$last_clean_backup" ] ; then
				resume_date=$b
				break
			fi
		done
	fi

	# create backup folder
	mkdir -p "$destination/$backup_date/$path_dest"
	if [ $? != 0 ] ; then
		internal_error "cannot create backup destination"
		srv_clean_exit 204
	fi

	# resume from old backup
	if lb_istrue $resume ; then
		log_debug "resume backup: move backup from $resume_date to $backup_date for $path_dest"

		# clean current backup dir and move old in it
		[ -n "$resume_date" ] && \
		rmdir "$destination/$backup_date/$path_dest" && \
		move_backup $resume_date $backup_date "$path_dest" &> /dev/null
		if [ $? = 0 ] ; then
			# delete old infofile
			clean_empty_backup -i $resume_date &> /dev/null
		else
			internal_error "resume failed"
			srv_clean_exit 202
		fi
	else
		# no resume

		# if last backup defined, prepare versionning
		if [ -n "$last_clean_backup" ] ; then
			# if  mirror mode or no hard links, move destination
			if [ $keep_limit = 0 ] || ! lb_istrue $hard_links ; then

				log_debug "move backup from $last_clean_backup to $backup_date for $path_dest"

				# move old backup as current backup and update latest link
				rmdir "$destination/$backup_date/$path_dest" && \
				move_backup $last_clean_backup $backup_date "$path_dest" &> /dev/null && \
				create_latest_link &> /dev/null

				if [ $? != 0 ] ; then
					print_error --log "failed to prepare destination"
					srv_clean_exit 202
				fi
			fi
		fi
	fi

	# If keep_limit = 0 (mirror mode), we don't need to use versionning.
	# If no hard links, do not create trash
	if [ -n "$last_clean_backup" ] && [ $keep_limit != 0 ] && ! lb_istrue $hard_links; then
		log_debug "create trash: $last_clean_backup"

		mkdir -p "$destination/$last_clean_backup/$path_dest" &> /dev/null
		if [ $? != 0 ] ; then
			print_error --log "failed to prepare trash"
			srv_clean_exit 202
		fi
	fi

	# trash mode
	if lb_istrue $trash_mode ; then
		# create trash
		mkdir -p "$destination/trash/$path_dest" &> /dev/null
		if [ $? != 0 ] ; then
			print_error --log "failed to prepare trash"
			srv_clean_exit 202
		fi
	fi

	# run rsync command
	"$rsync_path" "$@"
	res=$?

	log_debug "rsync result: $res"

	# save rsync result in info file and delete temporary file
	lb_set_config -s src1 "$infofile" rsync_result $res

	# forward exit code
	[ $res != 0 ] && lb_exitcode=$res

	# if cancel, do not consider as cancelled backup
	# (must be silent for some OS)
	catch_kills srv_clean_exit &> /dev/null

	# clean empty trash and infofile
	clean_empty_backup -i $last_clean_backup "$path_dest" &> /dev/null

	# clean directory
	clean_empty_backup -i $backup_date "$path_dest" &> /dev/null

	if rsync_result $res ; then
		# create latest backup directory link
		create_latest_link &> /dev/null

		# rotate backups
		rotate_backups &> /dev/null
	fi

	srv_clean_exit
}


# Restore files
# Usage: t2bs_restore [OPTIONS]
t2bs_restore() {

	local no_lock=false

	# get command options
	while [ $# -gt 0 ] ; do
		case $1 in
			--t2b-nolock)
				no_lock=true
				;;

			*)
				break
				;;
		esac
		shift # load next argument
	done

	# get infos from token
	for i in "${token_infos[@]}" ; do
		case $(echo "$i" | cut -d= -f1) in
			date)
				backup_date=$(echo "$i" | cut -d= -f2-)
				;;
		esac
	done

	# verify if required variables are there
	if [ -z "$backup_date" ] ; then
		internal_error "failed to get infos from token"
		return 202
	fi

	# check rsync syntax
	srv_check_rsync_options "$@" || return 201

	# check lock
	if ! $no_lock ; then
		if current_lock -q ; then
			print_error "destination locked"
			return 205
		fi
	fi

	log_debug "current restore from $backup_date"

	# catch term signals (must be silent for some OS)
	catch_kills srv_cancel_exit &> /dev/null

	# run rsync command
	"$rsync_path" "$@"
	res=$?

	log_debug "rsync result: $res"

	srv_clean_exit
}


# Get backup history
# Usage: t2bs_history [OPTIONS] PATH
t2bs_history() {

	# default option values
	local history_opts=() file

	# get options
	while [ $# -gt 0 ] ; do
		case $1 in
			-a|--all)
				history_opts=(-a)
				;;
			*)
				break
				;;
		esac
		shift # load next argument
	done

	# missing arguments
	[ -z "$*" ] && return 201

	# test backup destination
	if ! prepare_destination &> /dev/null ; then
		print_error "destination not ready"
		return 204
	fi

	# get backup versions of this file
	file_history=($(get_backup_history "${history_opts[@]}" "$*"))

	# print backup versions
	[ ${#file_history[@]} = 0 ] || echo ${file_history[*]}
}


# Rotate
# Usage: t2bs_rotate [OPTIONS] [LIMIT]
t2bs_rotate() {

	local keep=$keep_limit

	# test if number or period has a valid syntax
	if [ $# -gt 0 ] ; then
		if lb_is_integer "$1" ; then
			if [ $1 -lt 0 ] ; then
				print_help
				return 201
			fi
		else
			if ! test_period "$1" ; then
				print_help
				return 201
			fi
		fi

		keep=$1
	fi

	# test backup destination
	if ! prepare_destination &> /dev/null ; then
		print_error "destination not ready"
		return 204
	fi

	rotate_backups $keep || return 205
}


# Run a simple rsync command
# Usage: t2bs_rsync RSYNC_ARGS
t2bs_rsync() {
	# check rsync options
	srv_check_rsync_options "$@" || return 201

	# run rsync command
	"$rsync_path" "$@"
}


inc/functions-server.sh
#
#  time2backup server commands
#
#  This file is part of time2backup server (https://github.com/time2backup/server)
#
#  MIT License
#  Copyright (c) 2017-2021 Jean Prunneaux
#

# Index
#
#   Global functions
#     print_error
#     internal_error
#     usage_error
#     log_info
#     log_error
#     log_debug
#     debug
#   Security functions
#     srv_check_config
#     srv_save_token
#     srv_check_password
#     srv_check_token
#     srv_clean_tokens
#     srv_check_rsync_options
#   Backup steps
#     srv_clean_exit
#     srv_cancel_exit


#
#  Global functions
#

# Print an error to client
# Usage: print_error [--log] TEXT
print_error() {
	if [ "$1" = "--log" ] ; then
		shift
		log_error "$*"
	fi
	lb_error "t2b-server: $*"
}


# Print and return for an internal error
# Usage: internal_error [DETAILS]
internal_error() {
	print_error "internal server error"
	[ $# -gt 0 ] && log_error "$*"
	return 202
}


# Print an usage error
# Usage: usage_error [DETAILS]
usage_error() {
	print_error "bad command"
	[ $# -gt 0 ] && log_debug "bad command: $*"
	return 201
}


# Log info in server log
# Usage: log_info TEXT
log_info() {
	lb_log_info -d "[pid $$]" "$@"
}


# Log error in server log
# Usage: log_error TEXT
log_error() {
	lb_log_error -d "[pid $$]" "$@"
}


# Log debug in server log
# Usage: log_debug TEXT
log_debug() {
	lb_istrue $debug_mode || return 0
	lb_log_debug -d "[pid $$]" "$@"
}


# Debug function to overwrite time2backup functions output behaviour
# We redirect diplay messages to log file.
# Usage: debug TEXT
debug() {
	log_debug "$@"
}


#
#  Security functions
#


# Check config
# Usage: srv_check_config
srv_check_config() {
	# test if destination is set (if rsync, not needed)
	if [ "$command" != rsync ] ; then
		[ -z "$destination" ] && return 1
	fi

	lb_is_integer $token_expiration && [ $token_expiration -gt 0 ]
}


# Create and secure credential file
# Usage: srv_save_token TOKEN [ARGS]
# Dependencies: $credentials
# Exit codes:
#   0: token saved
#   1: save error
srv_save_token() {

	# if file exists
	if [ -e "$credentials" ] ; then
		# test if is a file
		if ! [ -f "$credentials" ] ; then
			log_error "credential path is not a file"
			return 1
		fi
	else
		# create & secure file
		touch "$credentials" && chmod 600 "$credentials"
		if [ $? != 0 ] ; then
			log_error "cannot create credential file"
			return 1
		fi
	fi

	# test if file is readable
	if ! [ -r "$credentials" ] ; then
		log_error "credential file is not readable"
		return 1
	fi

	# test if file is writable
	if ! [ -w "$credentials" ] ; then
		log_error "credential file is not writable"
		return 1
	fi

	# write token in file
	lb_join "	" "$(date +%s)" "$@" | tee -a "$credentials" &> /dev/null
	if [ $? != 0 ] ; then
		log_error "cannot write in credential file"
		return 1
	fi

	return 0
}


# Check client password
# Usage: srv_check_password [USER:]PASSWORD
srv_check_password() {
	# load authentication file if exists
	[ -f config/auth.conf ] || return 0

	# if passwords are set, test it
	if [ -s config/auth.conf ] ; then
		if ! grep -q "^$1$" config/auth.conf ; then
			print_error "authentication failed"
			log_error "Invalid password for $lb_current_user $ssh_info"
			return 1
		fi

		log_info "Authentification succeeded for $lb_current_user $ssh_info"
		return 0
	fi
}


# Check client token
# Usage: srv_check_token TOKEN
# Dependencies: $credentials
srv_check_token() {

	# search token in credentials file
	if [ -n "$1" ] ; then

		local session
		session=$(grep -E "^[0-9]+	$1	" "$credentials")

		if [ $? = 0 ] ; then
			log_info "Authentication by token successful for $lb_current_user $ssh_info"

			# get saved infos in credentials
			lb_split "	" "$session"
			token_infos=("${lb_split[@]}")
			return 0
		fi
	fi

	print_error "authentication failed"
	log_info "Invalid token for $lb_current_user $ssh_info"
	return 1
}


# Clean old tokens
# Usage: srv_clean_tokens
# Dependencies: $credentials
srv_clean_tokens() {
	# default expiration time
	if ! lb_is_integer "$token_expiration" || [ $token_expiration -le 10 ] ; then
		token_expiration=300
	fi

	for t in $(cat "$credentials" 2> /dev/null | awk '{print $1}') ; do
		# remove tokens older than 5 minutes
		if lb_is_integer $t && [ $(($t + $token_expiration)) -lt "$(date +%s)" ] ; then
			lb_edit "/^$t/d" "$credentials" &> /dev/null
		fi
	done
}


# Check rsync options
# Usage: srv_check_rsync_options ARGS
srv_check_rsync_options() {
	lb_in_array --server "$@" || internal_error "rsync bad syntax: $*"
}


#
#  Backup steps
#

# Clean & exit
# Usage: srv_clean_exit [EXIT_CODE]
srv_clean_exit() {
	# set exit code if specified
	[ -n "$1" ] && lb_exitcode=$1

	# delete backup lock
	release_lock &> /dev/null

	if [ "$command" = backup ] ; then
		clean_empty_backup -i $backup_date "$path_dest" &> /dev/null
	fi

	lb_exit
}


# Cancel exit
# Usage: srv_cancel_exit
srv_cancel_exit() {
	srv_clean_exit 217
}


t2b-server.sh
#!/bin/bash
#
#  time2backup server
#
#  Website: https://github.com/time2backup/server
#  MIT License
#  Copyright (c) 2017-2021 Jean Prunneaux
#
#  Version 1.0.3 (2021-03-04)
#

declare -r version=1.0.3


#
#  Initialization
#

# get real path of the script
if [ "$(uname)" = Darwin ] ; then
	# macOS which does not support readlink -f option
	current_script=$(perl -e 'use Cwd "abs_path";print abs_path(shift)' "$0")
else
	current_script=$(readlink -f "$0")
fi

# get directory of the current script
script_directory=$(dirname "$current_script")

cd "$script_directory"
if [ $? != 0 ] ; then
	echo >&2 "t2b-server: internal server error"
	exit 202
fi

# load libbash with GUI (because of t2b functions using it)
source libbash/libbash.sh -g 2> /dev/null
if [ $? != 0 ] ; then
	echo >&2 "t2b-server: internal server error"
	exit 202
fi

# force libbash GUI to console mode
lbg_set_gui console
notifications=false

# disable display messages
lb_set_display_level ERROR

# change current script name
lb_current_script_name=time2backup-server


#
#  Functions
#

# load init config
source inc/init.sh > /dev/null
if [ $? != 0 ] ; then
	lb_error "t2b-server: internal server error"
	exit 202
fi

# load time2backup functions
source inc/functions.sh > /dev/null
if [ $? != 0 ] ; then
	lb_error "t2b-server: internal server error"
	exit 202
fi

# load server functions
source inc/functions-server.sh > /dev/null
if [ $? != 0 ] ; then
	lb_error "t2b-server: internal server error"
	exit 202
fi

# load commands
source inc/commands.sh > /dev/null
if [ $? != 0 ] ; then
	lb_error "t2b-server: internal server error"
	exit 202
fi


#
#  Main program
#

# load the default config
if ! lb_import_config config/time2backup-server.example.conf ; then
	print_error --log "error in default config"
	exit 202
fi

# load config if exists
if [ -f config/time2backup-server.conf ] ; then
	# load config (securely)
	if ! lb_import_config -t config/time2backup-server.example.conf config/time2backup-server.conf ; then
		print_error --log "error in config"
		exit 202
	fi
fi

# set log level
lb_istrue $debug_mode || lb_set_log_level INFO

# get current context info
ssh_info=$SSH_CLIENT
[ -n "$ssh_info" ] && ssh_info="from $ssh_info"

# get global options
while [ $# -gt 0 ] ; do
	case $1 in
		-p)
			# password
			if [ -z "$2" ] ; then
				usage_error "global -p: missing password"
				exit 201
			fi
			client_password=$2
			shift
			;;

		-t)
			# token
			if [ -z "$2" ] ; then
				usage_error "global -t: missing token"
				exit 201
			fi
			token=$2
			shift
			;;

		*)
			break
			;;
	esac
	shift # load next argument
done

# get main command
if [ $# = 0 ] ; then
	usage_error "no main command"
	exit 201
fi

command=$1
shift

# test config
if ! srv_check_config ; then
	print_error --log "error in config"
	exit 202
fi

# test hard links compatibility
if ! lb_istrue $force_hard_links ; then
	test_hardlinks "$destination" || hard_links=false
fi

# reset default rsync path
[ -z "$rsync_path" ] && rsync_path=rsync

# set log file
[ -z "$logfile" ] && logfile=server.log
lb_set_logfile -a "$logfile"

# clean old tokens
srv_clean_tokens

# command operations
case $command in
	prepare|history|rotate|rsync)
		# check password
		srv_check_password "$client_password" || exit 203
		;;

	backup|restore)
		# check token
		srv_check_token "$token" || exit 203
		;;

	*)
		# bad command
		usage_error "$*"
		exit 201
		;;
esac

# run command
t2bs_$command "$@"
